// @remotion/renderer/dist/make-page.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makePage = void 0;
const puppeteer_evaluate_1 = require("./puppeteer-evaluate");
const set_props_and_env_1 = require("./set-props-and-env");
const makePage = async ({ context, initialFrame, browserReplacer, logLevel, indent, pagesArray, onBrowserLog, scale, timeoutInMilliseconds, composition, proxyPort, serveUrl, muted, envVariables, serializedInputPropsWithCustomSchema, imageFormat, serializedResolvedPropsWithCustomSchema, pageIndex, }) => {
    const page = await browserReplacer
        .getBrowser()
        .newPage({ context, logLevel, indent, pageIndex });
    pagesArray.push(page);
    await page.setViewport({
        width: composition.width,
        height: composition.height,
        deviceScaleFactor: scale,
    });
    const logCallback = (log) => {
        onBrowserLog === null || onBrowserLog === void 0 ? void 0 : onBrowserLog({
            stackTrace: log.stackTrace(),
            text: log.text,
            type: log.type,
        });
    };
    if (onBrowserLog) {
        page.on('console', logCallback);
    }
    await (0, set_props_and_env_1.setPropsAndEnv)({
        serializedInputPropsWithCustomSchema,
        envVariables,
        page,
        serveUrl,
        initialFrame,
        timeoutInMilliseconds,
        proxyPort,
        retriesRemaining: 2,
        audioEnabled: !muted,
        videoEnabled: imageFormat !== 'none',
        indent,
        logLevel,
        onServeUrlVisited: () => undefined,
    });
    await (0, puppeteer_evaluate_1.puppeteerEvaluateWithCatch)({
        // eslint-disable-next-line max-params
        pageFunction: (id, props, durationInFrames, fps, height, width, defaultCodec) => {
            window.remotion_setBundleMode({
                type: 'composition',
                compositionName: id,
                serializedResolvedPropsWithSchema: props,
                compositionDurationInFrames: durationInFrames,
                compositionFps: fps,
                compositionHeight: height,
                compositionWidth: width,
                compositionDefaultCodec: defaultCodec,
            });
        },
        args: [
            composition.id,
            serializedResolvedPropsWithCustomSchema,
            composition.durationInFrames,
            composition.fps,
            composition.height,
            composition.width,
            composition.defaultCodec,
        ],
        frame: null,
        page,
        timeoutInMilliseconds,
    });
    page.off('console', logCallback);
    return page;
};
exports.makePage = makePage;
