// @remotion/media-parser/dist/containers/transport-stream/handle-aac-packet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleAacPacket = void 0;
const aac_codecprivate_1 = require("../../aac-codecprivate");
const convert_audio_or_video_sample_1 = require("../../convert-audio-or-video-sample");
const register_track_1 = require("../../register-track");
const adts_header_1 = require("./adts-header");
const handle_avc_packet_1 = require("./handle-avc-packet");
const handleAacPacket = async ({ streamBuffer, state, programId, offset, }) => {
    var _a;
    const adtsHeader = (0, adts_header_1.readAdtsHeader)(streamBuffer.buffer);
    if (!adtsHeader) {
        throw new Error('Invalid ADTS header - too short');
    }
    const { channelConfiguration, codecPrivate, sampleRate, audioObjectType } = adtsHeader;
    const isTrackRegistered = state.callbacks.tracks.getTracks().find((t) => {
        return t.trackId === programId;
    });
    if (!isTrackRegistered) {
        const track = {
            type: 'audio',
            codecPrivate,
            trackId: programId,
            trakBox: null,
            timescale: handle_avc_packet_1.MPEG_TIMESCALE,
            codecWithoutConfig: 'aac',
            codec: (0, aac_codecprivate_1.mapAudioObjectTypeToCodecString)(audioObjectType),
            // https://www.w3.org/TR/webcodecs-aac-codec-registration/
            description: undefined,
            numberOfChannels: channelConfiguration,
            sampleRate,
        };
        await (0, register_track_1.registerTrack)({
            track,
            state,
            container: 'transport-stream',
        });
    }
    const sample = {
        cts: streamBuffer.pesHeader.pts,
        dts: (_a = streamBuffer.pesHeader.dts) !== null && _a !== void 0 ? _a : streamBuffer.pesHeader.pts,
        timestamp: streamBuffer.pesHeader.pts,
        duration: undefined,
        data: new Uint8Array(streamBuffer.buffer),
        trackId: programId,
        type: 'key',
        offset,
        timescale: handle_avc_packet_1.MPEG_TIMESCALE,
    };
    await state.callbacks.onAudioSample(programId, (0, convert_audio_or_video_sample_1.convertAudioOrVideoSampleToWebCodecsTimestamps)(sample, handle_avc_packet_1.MPEG_TIMESCALE));
};
exports.handleAacPacket = handleAacPacket;
