// @remotion/media-parser/dist/readers/from-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeReader = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const stream_1 = require("stream");
exports.nodeReader = {
    read: ({ src, range, controller }) => {
        if (typeof src !== 'string') {
            throw new Error('src must be a string when using `nodeReader`');
        }
        const ownController = new AbortController();
        const stream = (0, fs_1.createReadStream)(src, {
            start: range === null ? 0 : typeof range === 'number' ? range : range[0],
            end: range === null
                ? Infinity
                : typeof range === 'number'
                    ? Infinity
                    : range[1],
            signal: ownController.signal,
        });
        controller._internals.signal.addEventListener('abort', () => {
            ownController.abort();
        }, { once: true });
        const stats = (0, fs_1.statSync)(src);
        const reader = stream_1.Readable.toWeb(stream).getReader();
        if (controller) {
            controller._internals.signal.addEventListener('abort', () => {
                reader.cancel().catch(() => { });
            }, { once: true });
        }
        return Promise.resolve({
            reader: {
                reader,
                abort: () => {
                    ownController.abort();
                },
            },
            contentLength: stats.size,
            contentType: null,
            name: src.split(path_1.sep).pop(),
            supportsContentRange: true,
        });
    },
    getLength: (src) => {
        if (typeof src !== 'string') {
            throw new Error('src must be a string when using `nodeReader`');
        }
        const stats = (0, fs_1.statSync)(src);
        return Promise.resolve(stats.size);
    },
};
