// @remotion/renderer/dist/compositor/compose.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.callCompositor = exports.serializeCommand = void 0;
const node_child_process_1 = require("node:child_process");
const node_path_1 = __importDefault(require("node:path"));
const get_executable_path_1 = require("./get-executable-path");
const make_file_executable_1 = require("./make-file-executable");
const make_nonce_1 = require("./make-nonce");
const serializeCommand = (command, params) => {
    return {
        nonce: (0, make_nonce_1.makeNonce)(),
        payload: {
            type: command,
            params,
        },
    };
};
exports.serializeCommand = serializeCommand;
const callCompositor = (payload, indent, logLevel, binariesDirectory) => {
    return new Promise((resolve, reject) => {
        const execPath = (0, get_executable_path_1.getExecutablePath)({
            type: 'compositor',
            indent,
            logLevel,
            binariesDirectory,
        });
        (0, make_file_executable_1.makeFileExecutableIfItIsNot)(execPath);
        const child = (0, node_child_process_1.spawn)(execPath, [payload], {
            cwd: node_path_1.default.dirname(execPath),
        });
        const stderrChunks = [];
        child.stderr.on('data', (d) => stderrChunks.push(d));
        child.on('close', (code) => {
            if (code === 0) {
                resolve();
            }
            else {
                const message = Buffer.concat(stderrChunks).toString('utf-8');
                try {
                    // Try to see if the error is a JSON
                    const parsed = JSON.parse(message);
                    const msg = `Compositor error: ${parsed.error}`;
                    const err = new Error(`${msg}\n${parsed.backtrace}`);
                    reject(err);
                }
                catch (_a) {
                    reject(new Error(`Compositor panicked with code ${code}: ${message}`));
                }
            }
        });
        if (child.stdin.closed) {
            reject(new Error('Compositor stdin closed unexpectedly,' +
                Buffer.concat(stderrChunks).toString('utf-8')));
            return;
        }
        try {
            child.stdin.write(payload, (e) => {
                if (e) {
                    if (e instanceof Error && e.message.includes('EPIPE')) {
                        reject(new Error('Compositor stdin closed unexpectedly,' +
                            Buffer.concat(stderrChunks).toString('utf-8')));
                    }
                    return;
                }
                child.stdin.end();
            });
        }
        catch (err) {
            if (err instanceof Error && err.message.includes('EPIPE')) {
                reject(new Error('Compositor stdin closed unexpectedly,' +
                    Buffer.concat(stderrChunks).toString('utf-8')));
            }
        }
    });
};
exports.callCompositor = callCompositor;
