// @remotion/media-parser/dist/containers/mp3/get-duration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDurationFromMp3 = void 0;
const get_frame_length_1 = require("./get-frame-length");
const samples_per_mpeg_file_1 = require("./samples-per-mpeg-file");
const getDurationFromMp3 = (state) => {
    const mp3Info = state.mp3Info.getMp3Info();
    if (!mp3Info) {
        throw new Error('No mp3 info');
    }
    const samplesPerFrame = (0, samples_per_mpeg_file_1.getSamplesPerMpegFrame)({
        layer: mp3Info.layer,
        mpegVersion: mp3Info.mpegVersion,
    });
    const frameLengthInBytes = (0, get_frame_length_1.getMpegFrameLength)({
        bitrateKbit: mp3Info.bitrateKbit,
        padding: false,
        samplesPerFrame,
        samplingFrequency: mp3Info.sampleRate,
        layer: mp3Info.layer,
    });
    const frames = Math.floor((state.contentLength - mp3Info.startOfMpegStream) / frameLengthInBytes);
    const samples = frames * samplesPerFrame;
    const durationInSeconds = samples / mp3Info.sampleRate;
    return durationInSeconds;
};
exports.getDurationFromMp3 = getDurationFromMp3;
