// @remotion/media-parser/dist/containers/riff/expect-riff-box.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expectRiffBox = void 0;
const register_track_1 = require("../../register-track");
const get_tracks_from_avi_1 = require("./get-tracks-from-avi");
const is_movi_1 = require("./is-movi");
const parse_riff_box_1 = require("./parse-riff-box");
const expectRiffBox = async (state) => {
    const { iterator } = state;
    // Need at least 16 bytes to read LIST,size,movi,size
    if (state.iterator.bytesRemaining() < 16) {
        return null;
    }
    const checkpoint = iterator.startCheckpoint();
    const ckId = iterator.getByteString(4, false);
    const ckSize = iterator.getUint32Le();
    if ((0, is_movi_1.isMoviAtom)(iterator, ckId)) {
        iterator.discard(4);
        state.videoSection.setVideoSection({
            start: iterator.counter.getOffset(),
            size: ckSize - 4,
        });
        return null;
    }
    if (iterator.bytesRemaining() < ckSize) {
        checkpoint.returnToCheckpoint();
        return null;
    }
    const box = await (0, parse_riff_box_1.parseRiffBox)({
        id: ckId,
        size: ckSize,
        state,
    });
    if (box.type === 'strh-box') {
        if (box.strf.type === 'strf-box-audio' && state.onAudioTrack) {
            const audioTrack = (0, get_tracks_from_avi_1.makeAviAudioTrack)({
                index: state.riff.getNextTrackIndex(),
                strf: box.strf,
            });
            await (0, register_track_1.registerTrack)({
                state,
                track: audioTrack,
                container: 'avi',
            });
        }
        if (state.onVideoTrack && box.strf.type === 'strf-box-video') {
            const videoTrack = (0, get_tracks_from_avi_1.makeAviVideoTrack)({
                strh: box,
                index: state.riff.getNextTrackIndex(),
                strf: box.strf,
            });
            (0, register_track_1.registerVideoTrackWhenProfileIsAvailable)({
                state,
                track: videoTrack,
                container: 'avi',
            });
        }
        state.riff.incrementNextTrackIndex();
    }
    return box;
};
exports.expectRiffBox = expectRiffBox;
