// @remotion/media-parser/dist/download-and-parse-media.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadAndParseMedia = void 0;
const internal_parse_media_1 = require("./internal-parse-media");
const log_1 = require("./log");
const from_fetch_1 = require("./readers/from-fetch");
const downloadAndParseMedia = async (options) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8;
    const logLevel = (_a = options.logLevel) !== null && _a !== void 0 ? _a : 'info';
    const content = await options.writer.createContent({
        filename: 'hmm',
        mimeType: 'shouldnotmatter',
        logLevel,
    });
    const returnValue = await (0, internal_parse_media_1.internalParseMedia)({
        fields: (_b = options.fields) !== null && _b !== void 0 ? _b : null,
        logLevel,
        mode: 'download',
        onAudioCodec: (_c = options.onAudioCodec) !== null && _c !== void 0 ? _c : null,
        onAudioTrack: null,
        onContainer: (_d = options.onContainer) !== null && _d !== void 0 ? _d : null,
        onDimensions: (_e = options.onDimensions) !== null && _e !== void 0 ? _e : null,
        onDiscardedData: async (data) => {
            await content.write(data);
        },
        onDurationInSeconds: (_f = options.onDurationInSeconds) !== null && _f !== void 0 ? _f : null,
        onFps: (_g = options.onFps) !== null && _g !== void 0 ? _g : null,
        onImages: (_h = options.onImages) !== null && _h !== void 0 ? _h : null,
        onInternalStats: (_j = options.onInternalStats) !== null && _j !== void 0 ? _j : null,
        onIsHdr: (_k = options.onIsHdr) !== null && _k !== void 0 ? _k : null,
        onKeyframes: (_l = options.onKeyframes) !== null && _l !== void 0 ? _l : null,
        onLocation: (_m = options.onLocation) !== null && _m !== void 0 ? _m : null,
        onMetadata: (_o = options.onMetadata) !== null && _o !== void 0 ? _o : null,
        onMimeType: (_p = options.onMimeType) !== null && _p !== void 0 ? _p : null,
        onName: (_q = options.onName) !== null && _q !== void 0 ? _q : null,
        onNumberOfAudioChannels: (_r = options.onNumberOfAudioChannels) !== null && _r !== void 0 ? _r : null,
        onParseProgress: (_s = options.onParseProgress) !== null && _s !== void 0 ? _s : null,
        onRotation: (_t = options.onRotation) !== null && _t !== void 0 ? _t : null,
        onSampleRate: (_u = options.onSampleRate) !== null && _u !== void 0 ? _u : null,
        onSize: (_v = options.onSize) !== null && _v !== void 0 ? _v : null,
        onSlowAudioBitrate: (_w = options.onSlowAudioBitrate) !== null && _w !== void 0 ? _w : null,
        onSlowDurationInSeconds: (_x = options.onSlowDurationInSeconds) !== null && _x !== void 0 ? _x : null,
        onSlowFps: (_y = options.onSlowFps) !== null && _y !== void 0 ? _y : null,
        onSlowKeyframes: (_z = options.onSlowKeyframes) !== null && _z !== void 0 ? _z : null,
        onSlowNumberOfFrames: (_0 = options.onSlowNumberOfFrames) !== null && _0 !== void 0 ? _0 : null,
        onSlowVideoBitrate: (_1 = options.onSlowVideoBitrate) !== null && _1 !== void 0 ? _1 : null,
        onStructure: (_2 = options.onStructure) !== null && _2 !== void 0 ? _2 : null,
        onTracks: (_3 = options.onTracks) !== null && _3 !== void 0 ? _3 : null,
        onUnrotatedDimensions: (_4 = options.onUnrotatedDimensions) !== null && _4 !== void 0 ? _4 : null,
        onVideoCodec: (_5 = options.onVideoCodec) !== null && _5 !== void 0 ? _5 : null,
        onVideoTrack: null,
        progressIntervalInMs: (_6 = options.progressIntervalInMs) !== null && _6 !== void 0 ? _6 : null,
        reader: (_7 = options.reader) !== null && _7 !== void 0 ? _7 : from_fetch_1.fetchReader,
        controller: (_8 = options.controller) !== null && _8 !== void 0 ? _8 : undefined,
        src: options.src,
        onError: async (err) => {
            var _a, _b;
            const action = (_b = (await ((_a = options.onError) === null || _a === void 0 ? void 0 : _a.call(options, err)))) !== null && _b !== void 0 ? _b : { action: 'fail' };
            if (action.action === 'fail') {
                log_1.Log.verbose(logLevel, 'Removing content');
                await content.finish();
                await content.remove();
            }
            return action;
        },
        acknowledgeRemotionLicense: Boolean(options.acknowledgeRemotionLicense),
        apiName: 'parseAndDownloadMedia()',
    });
    await content.finish();
    return returnValue;
};
exports.downloadAndParseMedia = downloadAndParseMedia;
