// @remotion/media-parser/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasBeenAborted = exports.MediaParserAbortError = exports.IsAnUnsupportedAudioTypeError = exports.IsAnUnsupportedFileTypeError = exports.IsAPdfError = exports.IsAnImageError = exports.IsAGifError = void 0;
class IsAGifError extends Error {
    constructor({ message, mimeType, sizeInBytes, fileName, }) {
        super(message);
        this.fileName = 'IsAGifError';
        this.mimeType = mimeType;
        this.sizeInBytes = sizeInBytes;
        this.fileName = fileName;
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, IsAGifError);
        }
    }
}
exports.IsAGifError = IsAGifError;
class IsAnImageError extends Error {
    constructor({ dimensions, imageType, message, mimeType, sizeInBytes, fileName, }) {
        super(message);
        this.name = 'IsAnImageError';
        this.imageType = imageType;
        this.dimensions = dimensions;
        this.mimeType = mimeType;
        this.sizeInBytes = sizeInBytes;
        this.fileName = fileName;
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, IsAnImageError);
        }
    }
}
exports.IsAnImageError = IsAnImageError;
class IsAPdfError extends Error {
    constructor({ message, mimeType, sizeInBytes, fileName, }) {
        super(message);
        this.name = 'IsAPdfError';
        this.mimeType = mimeType;
        this.sizeInBytes = sizeInBytes;
        this.fileName = fileName;
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, IsAPdfError);
        }
    }
}
exports.IsAPdfError = IsAPdfError;
class IsAnUnsupportedFileTypeError extends Error {
    constructor({ message, mimeType, sizeInBytes, fileName, }) {
        super(message);
        this.name = 'IsAnUnsupportedFileTypeError';
        this.mimeType = mimeType;
        this.sizeInBytes = sizeInBytes;
        this.fileName = fileName;
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, IsAnUnsupportedFileTypeError);
        }
    }
}
exports.IsAnUnsupportedFileTypeError = IsAnUnsupportedFileTypeError;
class IsAnUnsupportedAudioTypeError extends Error {
    constructor({ message, mimeType, sizeInBytes, fileName, audioType, }) {
        super(message);
        this.name = 'IsAnUnsupportedAudioTypeError';
        this.mimeType = mimeType;
        this.sizeInBytes = sizeInBytes;
        this.fileName = fileName;
        this.audioType = audioType;
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, IsAnUnsupportedAudioTypeError);
        }
    }
}
exports.IsAnUnsupportedAudioTypeError = IsAnUnsupportedAudioTypeError;
class MediaParserAbortError extends Error {
    constructor(message) {
        super(message);
        this.name = 'MediaParserAbortError';
        this.cause = undefined;
    }
}
exports.MediaParserAbortError = MediaParserAbortError;
const hasBeenAborted = (error) => {
    return error instanceof MediaParserAbortError;
};
exports.hasBeenAborted = hasBeenAborted;
