// @remotion/media-parser/dist/media-parser-controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mediaParserController = void 0;
const emitter_1 = require("./emitter");
const pause_signal_1 = require("./pause-signal");
const mediaParserController = () => {
    const abortController = new AbortController();
    const emitter = new emitter_1.MediaParserEmitter();
    const pauseSignal = (0, pause_signal_1.makePauseSignal)(emitter);
    const checkForAbortAndPause = async () => {
        if (abortController.signal.aborted) {
            throw new Error('Aborted');
        }
        await pauseSignal.waitUntilResume();
    };
    return {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        abort: (reason) => {
            abortController.abort(reason);
        },
        pause: pauseSignal.pause,
        resume: pauseSignal.resume,
        addEventListener: emitter.addEventListener,
        removeEventListener: emitter.removeEventListener,
        _internals: {
            signal: abortController.signal,
            checkForAbortAndPause,
        },
    };
};
exports.mediaParserController = mediaParserController;
