// @remotion/media-parser/dist/state/need-samples-for-fields.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.needsToIterateOverSamples = void 0;
const needsSamples = {
    slowDurationInSeconds: true,
    slowFps: true,
    slowKeyframes: true,
    slowNumberOfFrames: true,
    audioCodec: false,
    container: false,
    dimensions: false,
    durationInSeconds: false,
    fps: false,
    internalStats: false,
    isHdr: false,
    name: false,
    rotation: false,
    size: false,
    structure: false,
    tracks: false,
    unrotatedDimensions: false,
    videoCodec: false,
    metadata: false,
    location: false,
    mimeType: false,
    keyframes: false,
    images: false,
    numberOfAudioChannels: false,
    sampleRate: false,
    slowAudioBitrate: true,
    slowVideoBitrate: true,
};
const needsToIterateOverSamples = ({ fields, emittedFields, }) => {
    const keys = Object.keys(fields !== null && fields !== void 0 ? fields : {});
    const selectedKeys = keys.filter((k) => fields[k]);
    return selectedKeys.some((k) => needsSamples[k] && !emittedFields[k]);
};
exports.needsToIterateOverSamples = needsToIterateOverSamples;
