// @remotion/media-parser/dist/state/parser-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeParserState = void 0;
const buffer_iterator_1 = require("../buffer-iterator");
const log_1 = require("../log");
const aac_state_1 = require("./aac-state");
const emitted_fields_1 = require("./emitted-fields");
const flac_state_1 = require("./flac-state");
const images_1 = require("./images");
const iso_state_1 = require("./iso-base-media/iso-state");
const keyframes_1 = require("./keyframes");
const last_eventloop_break_1 = require("./last-eventloop-break");
const mp3_1 = require("./mp3");
const riff_1 = require("./riff");
const sample_callbacks_1 = require("./sample-callbacks");
const slow_duration_fps_1 = require("./slow-duration-fps");
const structure_1 = require("./structure");
const transport_stream_1 = require("./transport-stream");
const video_section_1 = require("./video-section");
const webm_1 = require("./webm");
const makeParserState = ({ hasAudioTrackHandlers, hasVideoTrackHandlers, controller, fields, onAudioTrack, onVideoTrack, contentLength, logLevel, mode, src, readerInterface, onDiscardedData, }) => {
    let skippedBytes = 0;
    const iterator = (0, buffer_iterator_1.getArrayBufferIterator)(new Uint8Array([]), contentLength);
    const increaseSkippedBytes = (bytes) => {
        skippedBytes += bytes;
    };
    const structure = (0, structure_1.structureState)();
    const keyframes = (0, keyframes_1.keyframesState)();
    const emittedFields = (0, emitted_fields_1.emittedState)();
    const slowDurationAndFps = (0, slow_duration_fps_1.slowDurationAndFpsState)();
    const mp3Info = (0, mp3_1.makeMp3State)();
    const images = (0, images_1.imagesState)();
    const discardReadBytes = async (force) => {
        const { bytesRemoved, removedData } = iterator.removeBytesRead(force, mode);
        if (bytesRemoved) {
            log_1.Log.verbose(logLevel, `Freed ${bytesRemoved} bytes`);
        }
        if (removedData && onDiscardedData) {
            await onDiscardedData(removedData);
        }
    };
    return {
        riff: (0, riff_1.riffSpecificState)(),
        transportStream: (0, transport_stream_1.transportStreamState)(),
        webm: (0, webm_1.webmState)(),
        iso: (0, iso_state_1.isoBaseMediaState)(),
        mp3Info,
        aac: (0, aac_state_1.aacState)(),
        flac: (0, flac_state_1.flacState)(),
        callbacks: (0, sample_callbacks_1.sampleCallback)({
            controller,
            hasAudioTrackHandlers,
            hasVideoTrackHandlers,
            fields,
            keyframes,
            emittedFields,
            slowDurationAndFpsState: slowDurationAndFps,
            structure,
        }),
        getInternalStats: () => {
            var _a;
            return ({
                skippedBytes,
                finalCursorOffset: (_a = iterator.counter.getOffset()) !== null && _a !== void 0 ? _a : 0,
            });
        },
        getSkipBytes: () => skippedBytes,
        increaseSkippedBytes,
        keyframes,
        ...structure,
        onAudioTrack,
        onVideoTrack,
        emittedFields,
        fields,
        slowDurationAndFps,
        contentLength,
        images,
        videoSection: (0, video_section_1.videoSectionState)(),
        logLevel,
        iterator,
        controller,
        mode,
        eventLoop: (0, last_eventloop_break_1.eventLoopState)(logLevel),
        src,
        readerInterface,
        discardReadBytes,
    };
};
exports.makeParserState = makeParserState;
