// @remotion/media-parser/dist/state/slow-duration-fps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.slowDurationAndFpsState = void 0;
const slowDurationAndFpsState = () => {
    let smallestVideoSample;
    let largestVideoSample;
    let smallestAudioSample;
    let largestAudioSample;
    let audioSizesInBytes = 0;
    let videoSizeInBytes = 0;
    let videoSamples = 0;
    let audioSamples = 0;
    const getSlowVideoDurationInSeconds = () => {
        let videoDuration = null;
        if (smallestVideoSample !== undefined && largestVideoSample !== undefined) {
            const startingTimestampDifference = largestVideoSample - smallestVideoSample;
            const timeBetweenSamples = startingTimestampDifference / (videoSamples - 1);
            videoDuration = timeBetweenSamples * videoSamples;
        }
        return videoDuration;
    };
    const getSlowDurationInSeconds = () => {
        const videoDuration = getSlowVideoDurationInSeconds();
        let audioDuration = null;
        if (smallestAudioSample !== undefined && largestAudioSample !== undefined) {
            const startingTimestampDifferenceAudio = largestAudioSample - smallestAudioSample;
            const timeBetweenSamplesAudio = startingTimestampDifferenceAudio / (audioSamples - 1);
            audioDuration = timeBetweenSamplesAudio * audioSamples;
        }
        if (videoDuration === null && audioDuration === null) {
            throw new Error('No samples');
        }
        return Math.max(videoDuration !== null && videoDuration !== void 0 ? videoDuration : 0, audioDuration !== null && audioDuration !== void 0 ? audioDuration : 0);
    };
    return {
        addVideoSample: (videoSample) => {
            videoSamples++;
            const presentationTimeInSeconds = videoSample.cts / videoSample.timescale;
            if (largestVideoSample === undefined ||
                presentationTimeInSeconds > largestVideoSample) {
                largestVideoSample = presentationTimeInSeconds;
            }
            if (smallestVideoSample === undefined ||
                presentationTimeInSeconds < smallestVideoSample) {
                smallestVideoSample = presentationTimeInSeconds;
            }
            videoSizeInBytes += videoSample.data.byteLength;
        },
        addAudioSample: (audioSample) => {
            audioSamples++;
            const presentationTimeInSeconds = audioSample.cts / audioSample.timescale;
            if (largestAudioSample === undefined ||
                presentationTimeInSeconds > largestAudioSample) {
                largestAudioSample = presentationTimeInSeconds;
            }
            if (smallestAudioSample === undefined ||
                presentationTimeInSeconds < smallestAudioSample) {
                smallestAudioSample = presentationTimeInSeconds;
            }
            audioSizesInBytes += audioSample.data.byteLength;
        },
        getSlowDurationInSeconds,
        getFps: () => {
            var _a;
            const videoDuration = (_a = getSlowVideoDurationInSeconds()) !== null && _a !== void 0 ? _a : 0;
            if (videoDuration === 0) {
                return 0;
            }
            return videoSamples / videoDuration;
        },
        getSlowNumberOfFrames: () => videoSamples,
        getAudioBitrate: () => {
            const audioDuration = getSlowDurationInSeconds();
            if (audioDuration === 0 || audioSizesInBytes === 0) {
                return null;
            }
            return (audioSizesInBytes * 8) / audioDuration;
        },
        getVideoBitrate: () => {
            const videoDuration = getSlowDurationInSeconds();
            if (videoDuration === 0 || videoSizeInBytes === 0) {
                return null;
            }
            return (videoSizeInBytes * 8) / videoDuration;
        },
    };
};
exports.slowDurationAndFpsState = slowDurationAndFpsState;
