// @remotion/media-parser/dist/containers/iso-base-media/get-moov-atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMoovAtom = void 0;
const log_1 = require("../../log");
const register_track_1 = require("../../register-track");
const parser_state_1 = require("../../state/parser-state");
const process_box_1 = require("./process-box");
const getMoovAtom = async ({ endOfMdat, state, }) => {
    const start = Date.now();
    log_1.Log.verbose(state.logLevel, 'Starting second fetch to get moov atom');
    const { reader } = await state.readerInterface.read({
        src: state.src,
        range: endOfMdat,
        controller: state.controller,
    });
    const childState = (0, parser_state_1.makeParserState)({
        hasAudioTrackHandlers: false,
        hasVideoTrackHandlers: false,
        controller: state.controller,
        fields: {
            structure: true,
        },
        onAudioTrack: state.onAudioTrack
            ? async ({ track, container }) => {
                await (0, register_track_1.registerTrack)({ state, track, container });
                return null;
            }
            : null,
        onVideoTrack: state.onVideoTrack
            ? async ({ track, container }) => {
                await (0, register_track_1.registerTrack)({ state, track, container });
                return null;
            }
            : null,
        contentLength: state.contentLength,
        logLevel: state.logLevel,
        mode: 'query',
        readerInterface: state.readerInterface,
        src: state.src,
        onDiscardedData: null,
    });
    while (true) {
        const result = await reader.reader.read();
        if (result.value) {
            childState.iterator.addData(result.value);
        }
        if (result.done) {
            break;
        }
    }
    const boxes = [];
    while (true) {
        const box = await (0, process_box_1.processBox)(childState);
        if (box) {
            boxes.push(box);
        }
        if (childState.iterator.counter.getOffset() + endOfMdat >
            state.contentLength) {
            throw new Error('Read past end of file');
        }
        if (childState.iterator.counter.getOffset() + endOfMdat ===
            state.contentLength) {
            break;
        }
    }
    const moov = boxes.find((b) => b.type === 'moov-box');
    if (!moov) {
        throw new Error('No moov box found');
    }
    log_1.Log.verbose(state.logLevel, `Finished fetching moov atom in ${Date.now() - start}ms`);
    return moov;
};
exports.getMoovAtom = getMoovAtom;
