// @remotion/media-parser/dist/containers/iso-base-media/get-keyframes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getKeyframesFromIsoBaseMedia = void 0;
const get_tracks_1 = require("../../get-tracks");
const get_sample_positions_from_track_1 = require("./get-sample-positions-from-track");
const traversal_1 = require("./traversal");
const getKeyframesFromIsoBaseMedia = (state) => {
    const { videoTracks } = (0, get_tracks_1.getTracksFromIsoBaseMedia)(state);
    const structure = state.getIsoStructure();
    const moofBox = (0, traversal_1.getMoofBoxes)(structure.boxes);
    const allSamples = videoTracks.map((t) => {
        const { timescale: ts } = t;
        const samplePositions = (0, get_sample_positions_from_track_1.getSamplePositionsFromTrack)({
            trakBox: t.trakBox,
            moofBoxes: moofBox,
        });
        const keyframes = samplePositions
            .filter((k) => {
            return k.isKeyframe;
        })
            .map((k) => {
            return {
                trackId: t.trackId,
                presentationTimeInSeconds: k.cts / ts,
                decodingTimeInSeconds: k.dts / ts,
                positionInBytes: k.offset,
                sizeInBytes: k.size,
            };
        });
        return keyframes;
    });
    return allSamples.flat();
};
exports.getKeyframesFromIsoBaseMedia = getKeyframesFromIsoBaseMedia;
