// @remotion/media-parser/dist/containers/iso-base-media/traversal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTrunBoxes = exports.getTfhdBox = exports.getTfdtBox = exports.getStssBox = exports.getStscBox = exports.getStszBox = exports.getCttsBox = exports.getSttsBox = exports.getStcoBox = exports.getVideoDescriptors = exports.getStsdBox = exports.getStblBox = exports.getMdhdBox = exports.getMdiaBox = exports.getTkhdBox = exports.getTraks = exports.getMvhdBox = exports.getMoofBoxes = exports.getMoovBox = exports.getFtypBox = void 0;
const getFtypBox = (segments) => {
    const ftypBox = segments.find((s) => s.type === 'ftyp-box');
    if (!ftypBox || ftypBox.type !== 'ftyp-box') {
        return null;
    }
    return ftypBox;
};
exports.getFtypBox = getFtypBox;
const getMoovBox = (state) => {
    if (state.iso.moov.getMoovBox()) {
        return state.iso.moov.getMoovBox();
    }
    const structure = state.getIsoStructure();
    const moovBox = structure.boxes.find((s) => s.type === 'moov-box');
    if (!moovBox || moovBox.type !== 'moov-box') {
        return null;
    }
    return moovBox;
};
exports.getMoovBox = getMoovBox;
const getMoofBoxes = (main) => {
    const moofBoxes = main.filter((s) => s.type === 'regular-box' && s.boxType === 'moof');
    return moofBoxes;
};
exports.getMoofBoxes = getMoofBoxes;
const getMvhdBox = (moovBox) => {
    const mvHdBox = moovBox.children.find((s) => s.type === 'mvhd-box');
    if (!mvHdBox || mvHdBox.type !== 'mvhd-box') {
        return null;
    }
    return mvHdBox;
};
exports.getMvhdBox = getMvhdBox;
const getTraks = (moovBox) => {
    return moovBox.children.filter((s) => s.type === 'trak-box');
};
exports.getTraks = getTraks;
const getTkhdBox = (trakBox) => {
    const tkhdBox = trakBox.children.find((s) => s.type === 'tkhd-box');
    return tkhdBox;
};
exports.getTkhdBox = getTkhdBox;
const getMdiaBox = (trakBox) => {
    const mdiaBox = trakBox.children.find((s) => s.type === 'regular-box' && s.boxType === 'mdia');
    if (!mdiaBox || mdiaBox.type !== 'regular-box') {
        return null;
    }
    return mdiaBox;
};
exports.getMdiaBox = getMdiaBox;
const getMdhdBox = (trakBox) => {
    const mdiaBox = (0, exports.getMdiaBox)(trakBox);
    if (!mdiaBox) {
        return null;
    }
    const mdhdBox = mdiaBox.children.find((c) => c.type === 'mdhd-box');
    return mdhdBox;
};
exports.getMdhdBox = getMdhdBox;
const getStblBox = (trakBox) => {
    const mdiaBox = (0, exports.getMdiaBox)(trakBox);
    if (!mdiaBox) {
        return null;
    }
    const minfBox = mdiaBox.children.find((s) => s.type === 'regular-box' && s.boxType === 'minf');
    if (!minfBox || minfBox.type !== 'regular-box') {
        return null;
    }
    const stblBox = minfBox.children.find((s) => s.type === 'regular-box' && s.boxType === 'stbl');
    if (!stblBox || stblBox.type !== 'regular-box') {
        return null;
    }
    return stblBox;
};
exports.getStblBox = getStblBox;
const getStsdBox = (trakBox) => {
    const stblBox = (0, exports.getStblBox)(trakBox);
    if (!stblBox || stblBox.type !== 'regular-box') {
        return null;
    }
    const stsdBox = stblBox.children.find((s) => s.type === 'stsd-box');
    return stsdBox;
};
exports.getStsdBox = getStsdBox;
const getVideoDescriptors = (trakBox) => {
    var _a;
    const stsdBox = (0, exports.getStsdBox)(trakBox);
    if (!stsdBox) {
        return null;
    }
    const descriptors = stsdBox.samples.map((s) => {
        return s.type === 'video'
            ? s.descriptors.map((d) => {
                return d.type === 'avcc-box'
                    ? d.privateData
                    : d.type === 'hvcc-box'
                        ? d.privateData
                        : null;
            })
            : [];
    });
    return (_a = descriptors.flat(1).filter(Boolean)[0]) !== null && _a !== void 0 ? _a : null;
};
exports.getVideoDescriptors = getVideoDescriptors;
const getStcoBox = (trakBox) => {
    const stblBox = (0, exports.getStblBox)(trakBox);
    if (!stblBox || stblBox.type !== 'regular-box') {
        return null;
    }
    const stcoBox = stblBox.children.find((s) => s.type === 'stco-box');
    return stcoBox;
};
exports.getStcoBox = getStcoBox;
const getSttsBox = (trakBox) => {
    const stblBox = (0, exports.getStblBox)(trakBox);
    if (!stblBox || stblBox.type !== 'regular-box') {
        return null;
    }
    const sttsBox = stblBox.children.find((s) => s.type === 'stts-box');
    return sttsBox;
};
exports.getSttsBox = getSttsBox;
const getCttsBox = (trakBox) => {
    const stblBox = (0, exports.getStblBox)(trakBox);
    if (!stblBox || stblBox.type !== 'regular-box') {
        return null;
    }
    const cttsBox = stblBox.children.find((s) => s.type === 'ctts-box');
    return cttsBox;
};
exports.getCttsBox = getCttsBox;
const getStszBox = (trakBox) => {
    const stblBox = (0, exports.getStblBox)(trakBox);
    if (!stblBox || stblBox.type !== 'regular-box') {
        return null;
    }
    const stszBox = stblBox.children.find((s) => s.type === 'stsz-box');
    return stszBox;
};
exports.getStszBox = getStszBox;
const getStscBox = (trakBox) => {
    const stblBox = (0, exports.getStblBox)(trakBox);
    if (!stblBox || stblBox.type !== 'regular-box') {
        return null;
    }
    const stcoBox = stblBox.children.find((b) => b.type === 'stsc-box');
    return stcoBox;
};
exports.getStscBox = getStscBox;
const getStssBox = (trakBox) => {
    const stblBox = (0, exports.getStblBox)(trakBox);
    if (!stblBox || stblBox.type !== 'regular-box') {
        return null;
    }
    const stssBox = stblBox.children.find((b) => b.type === 'stss-box');
    return stssBox;
};
exports.getStssBox = getStssBox;
const getTfdtBox = (segment) => {
    if (segment.type !== 'regular-box' || segment.boxType !== 'traf') {
        throw new Error('Expected traf-box');
    }
    const tfhdBox = segment.children.find((c) => c.type === 'tfdt-box');
    if (!tfhdBox || tfhdBox.type !== 'tfdt-box') {
        throw new Error('Expected tfhd-box');
    }
    return tfhdBox;
};
exports.getTfdtBox = getTfdtBox;
const getTfhdBox = (segment) => {
    if (segment.type !== 'regular-box' || segment.boxType !== 'traf') {
        throw new Error('Expected traf-box');
    }
    const tfhdBox = segment.children.find((c) => c.type === 'tfhd-box');
    if (!tfhdBox || tfhdBox.type !== 'tfhd-box') {
        throw new Error('Expected tfhd-box');
    }
    return tfhdBox;
};
exports.getTfhdBox = getTfhdBox;
const getTrunBoxes = (segment) => {
    if (segment.type !== 'regular-box' || segment.boxType !== 'traf') {
        throw new Error('Expected traf-box');
    }
    const trunBoxes = segment.children.filter((c) => c.type === 'trun-box');
    return trunBoxes;
};
exports.getTrunBoxes = getTrunBoxes;
