// @remotion/media-parser/dist/state/iso-base-media/cached-sample-positions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cachedSamplePositionsState = exports.calculateFlatSamples = void 0;
const get_sample_positions_from_track_1 = require("../../containers/iso-base-media/get-sample-positions-from-track");
const traversal_1 = require("../../containers/iso-base-media/traversal");
const get_tracks_1 = require("../../get-tracks");
const calculateFlatSamples = (state) => {
    const tracks = (0, get_tracks_1.getTracks)(state);
    const allTracks = [
        ...tracks.videoTracks,
        ...tracks.audioTracks,
        ...tracks.otherTracks,
    ];
    const flatSamples = allTracks
        .map((track) => {
        const samplePositions = (0, get_sample_positions_from_track_1.getSamplePositionsFromTrack)({
            trakBox: track.trakBox,
            moofBoxes: (0, traversal_1.getMoofBoxes)(state.getIsoStructure().boxes),
        });
        if (!samplePositions) {
            throw new Error('No sample positions');
        }
        return samplePositions.map((samplePosition) => {
            return {
                track,
                samplePosition,
            };
        });
    })
        .flat(1);
    return flatSamples;
};
exports.calculateFlatSamples = calculateFlatSamples;
const cachedSamplePositionsState = () => {
    const cachedForMdatStar = {};
    return {
        getSamples: (mdatStart) => {
            if (cachedForMdatStar[mdatStart]) {
                return cachedForMdatStar[mdatStart];
            }
            return null;
        },
        setSamples: (mdatStart, samples) => {
            cachedForMdatStar[mdatStart] = samples;
        },
    };
};
exports.cachedSamplePositionsState = cachedSamplePositionsState;
